src/file_watcher/mod.rs

```diff
@@ -0,0 +1,74 @@
+use std::io::{Error, ErrorKind, Result};
+use std::path::PathBuf;
+use std::vec::Vec;
+
+extern crate inotify;
+use inotify::Inotify;
+
+pub struct FileWatcher {
+    inotify: Inotify,
+    watches: Vec<FileWatch>
+}
+
+pub struct FileWatch {
+    descriptor: inotify::WatchDescriptor,
+}
+
+impl FileWatcher {
+    pub fn init() -> Result<FileWatcher>  {
+        let ino = match Inotify::init() {
+            Ok(i) => i,
+            Err(msg) => return Result::Err(msg),
+        };
+
+        return Result::Ok(FileWatcher {
+            inotify: ino,
+            watches: vec![]
+        });
+    }
+
+    pub fn close(self) -> Result<()> {
+        return self.inotify.close();
+    }
+
+    pub fn add_watch(&mut self, file_path: &PathBuf) -> Result<&FileWatch> {
+        let mask: inotify::WatchMask = inotify::WatchMask::MODIFY;
+
+        let watch = match self.inotify.add_watch(file_path, mask) {
+            Ok(w) => w,
+            Err(msg) => return Result::Err(msg),
+        };
+
+        let fw = FileWatch {
+            descriptor: watch,
+        };
+
+        self.watches.push(fw);
+        return Result::Ok(&self.watches.last().unwrap());
+    }
+
+    pub fn rm_watch(&mut self, fw: &FileWatch) -> Result<()> {
+        for i in 0..self.watches.len() {
+            let item_ref = self.watches.get(i).unwrap();
+            if item_ref.descriptor == fw.descriptor {
+                let item = self.watches.remove(i);
+                return self.inotify.rm_watch(item.descriptor);
+            }
+        }
```

Comment:
Just so you know, there is another way how to implement this. To be absolutely clear, I like your way better than the one I'm going to write here, but I don't know if you're aware a pattern like this exist. It might be more useful and clear many times, but in this case, your implementation wins imo. This comment is not meant to be a request for change or something, just a notice that there is another way how to do this:

```suggestion
        // Here, we create tuple of `(index, item)` from the watches, and get the first one that has the same descriptor as `fw`
        let watch_to_remove = self.watches.iter().enumerate().skip_while(|w| w.1.descriptor != fw.descriptor).next();
        if let Some(watch_to_remove) = watch_to_remove {
            // We extract index from the tuple. This isn't necessary and this block could be one line shorter, but then compiler emits a warning that it might become hard error in the future.
            let watch_idx_to_remove = watch_to_remove.0;
            let removed_fw = self.watches.remove(watch_idx_to_remove);
            return self.inotify.rm_watch(removed_fw.descriptor);
        }

```